source/agora/common/Bech32.d

```diff
@@ -0,0 +1,232 @@
+/*******************************************************************************
+
+    Utilities for converting between bech32-encoded string and `ubyte` array.
+
+    Copyright:
+        Copyright (c) 2021 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.common.Bech32;
+
+import agora.crypto.ECC;
+
+/** The Bech32 character set for encoding. */
+const char* CHARSET = "qpzry9x8gf2tvdw0s3jn54khce6mua7l";
+
+/** The Bech32 character set for decoding. */
+const byte[128] CHARSET_REV = [
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    15, -1, 10, 17, 21, 20, 26, 30,  7,  5, -1, -1, -1, -1, -1, -1,
+    -1, 29, -1, 24, 13, 25,  9,  8, 23, -1, 18, 22, 31, 27, 19, -1,
+     1,  0,  3, 16, 11, 28, 12, 14,  6,  4,  2, -1, -1, -1, -1, -1,
+    -1, 29, -1, 24, 13, 25,  9,  8, 23, -1, 18, 22, 31, 27, 19, -1,
+     1,  0,  3, 16, 11, 28, 12, 14,  6,  4,  2, -1, -1, -1, -1, -1
+];
+
+/** Concatenate two byte arrays. */
+ubyte[] cat(const ubyte[] x, const ubyte[] y) {
+    ubyte[] combined;
+    combined ~= x;
+    combined ~= y;
+    return combined;
+}
+
+/** This function will compute what 6 5-bit values to XOR into the last 6 input values, in order to
+ *  make the checksum 0. These 6 values are packed together in a single 30-bit integer. The higher
+ *  bits correspond to earlier values. */
+uint polymod(const ubyte[] values)
+{
+    // The input is interpreted as a list of coefficients of a polynomial over F = GF(32), with an
```

Comment:
This was clearly taken from some other code. Please keep existing copyrights in place.

Reply:
I will. Thanks.